### ISSUE TITLE
PyTorch automatic mixed precision (AMP) support

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1334]
This issue is to track the state of automatic mixed precision (AMP) support in the PyTorch engine (#1120).

This uses `autocast` and `GradScaler`, although both are orthogonal to each other.

**Comment by albertz** [comment]
Implemented via #1322.

**Comment by albertz** [comment]
Note, one thing which was confusing to me initially:

The parameters are still stored in float32, and then casted in each minibatch iteration. I wondered why the parameters are not directly stored in float16 or whatever you want to use for AMP.

I was just checking the original paper introducing [automatic mixed precision training](https://arxiv.org/abs/1710.03740), and it explains it (Sec 3.1):

> In mixed precision training, weights, activations and gradients are stored as FP16. In order to match the accuracy of the FP32 networks, an FP32 master copy of weights is maintained and updated with the weight gradient during the optimizer step. In each iteration an FP16 copy of the master weights is used in the forward and backward pass. ...
>
> While the need for FP32 master weights is not universal, there are two possible reasons why a number of networks require it. One explanation is that updates (weight gradients multiplied by the learning rate) become too small to be represented in FP16 - any value whose magnitude is smaller than $2^{−24}$ becomes zero in FP16. ...
>
> Another explanation is that the ratio of the weight value to the weight update is very large. In this case, even though the weight update is representable in FP16, it could still become zero when addition operation right-shifts it to align the binary point with the weight. ...

**Comment by JackTemaki** [comment]
> The parameters are still stored in float32, and then casted in each minibatch iteration.

As I said before, I still think this is wrong. They are not casted in each minibatch iteration, but the fp16 values are cached.

**Comment by albertz** [comment]
> They are not casted in each minibatch iteration, but the fp16 values are cached

I think this is wrong. Where do you read that? You are updating the float32 params in the end, so in any case, you need to do a cast again from the f32 params to f16 at some point, for every minibatch iteration, or at least after every param update.

**Comment by JackTemaki** [comment]
Yes, the cache is only used when accessing the same weight twice within the same autocast environment.

I found the actual code e.g. here:
https://github.com/pytorch/pytorch/blob/4eaaa086231b911ef768dd2d959f837b082efee6/aten/src/ATen/autocast_mode.cpp#L167

**Comment by albertz** [comment]
I forgot to answer. Yes, this makes sense. This however is a rare case for us. We rarely use some weight twice in the same model. So, still, every mini-batch, you do: cast params from f32 to f16, do comp, calc grad (f16), Adam update (I assume already f32 again for the momentums), then param update (f32). If you do grad accumulation, maybe also in f32? Not sure. In any case, at some point it gets casted back from f16 to f32. And probably in some intermediate calculations as well, which are better in f32. I think any reduce_sum and ops like this.

Btw, regarding my initial question (why not keep the weights as f16), despite what I quoted above from the original paper, I just stumbled upon this new paper: [In Defense of Pure 16-bit Floating-Point Neural Networks](https://arxiv.org/abs/2305.10947). I'm not really convinced by their experiments though. Also, I did not really mean to use pure f16 for everything, I just meant it for the params (and maybe optimizer state).
